### PR TITLE
IGNITE-16988 [Native Persistence 3.0] Porting FileIO from 2.0

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-16988

### CODE REVIEW

File: modules/file-io/pom.xml [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        </dependency>
+
+        <dependency>
+            <groupId>org.hamcrest</groupId>
+            <artifactId>hamcrest</artifactId>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
+
+</project>
```
Comment (ibessonov) [inline modules/file-io/pom.xml:62]: Empty line forgotten

File: modules/file-io/src/test/java/org/apache/ignite/internal/fileio/AbstractFileIoTest.java [hunk 2]
```diff
@@ -0,0 +1,588 @@
...
+ */
+@ExtendWith(WorkDirectoryExtension.class)
+public abstract class AbstractFileIoTest {
+    protected FileIoFactory fileIoFactory;
+
+    @WorkDirectory
+    protected Path workDir;
+
+    @Test
+    abstract void testFileIoFactory() throws Exception;
```
Comment (ibessonov) [inline modules/file-io/src/test/java/org/apache/ignite/internal/fileio/AbstractFileIoTest.java:52]: Why does this exist?

File: modules/file-io/src/test/java/org/apache/ignite/internal/fileio/AbstractFileIoTest.java [hunk 3]
```diff
@@ -0,0 +1,588 @@
...
+
+    @Test
+    void testWrite() throws Exception {
+        byte[] randomBytes = randomByteArray(4 * 1024);
+
+        Path testFilePath = workDir.resolve("test");
+
+        FileIo fileIo = fileIoFactory.create(testFilePath);
+
+        assertEquals(1024, fileIo.write(ByteBuffer.wrap(copyOfRange(randomBytes, 0, 1024))));
```
Comment (ibessonov) [inline modules/file-io/src/test/java/org/apache/ignite/internal/fileio/AbstractFileIoTest.java:249]: Can we also test slices? Also, this piece of code can be extracted in to a method, like "ByteBuffer rangeBuffer(arr, from, to)"

File: modules/file-io/src/test/java/org/apache/ignite/internal/fileio/AbstractFileIoTest.java [hunk 4]
```diff
@@ -0,0 +1,588 @@
...
+
+        assertNotNull(mmap);
+
+        assertEquals(0, mmap.position());
+
+        assertEquals(1024, mmap.capacity());
+    }
+
+    @Test
+    void testForce() throws Exception {
```
Comment (ibessonov) [inline modules/file-io/src/test/java/org/apache/ignite/internal/fileio/AbstractFileIoTest.java:423]: This one is kinda pointless, but let's leave it. How do you even test that sync has actually been invoked? I've no idea!

File: modules/file-io/src/test/java/org/apache/ignite/internal/fileio/AbstractFileIoTest.java [hunk 5]
```diff
@@ -0,0 +1,588 @@
...
+        }
+
+        byte[] expectedBytes = new byte[1024];
+
+        System.arraycopy(randomByteArray, 0, expectedBytes, 512, 512);
+
+        assertArrayEquals(expectedBytes, toByteArray(test1FilePath));
+    }
+
+    // TODO: IGNITE-16988 continue write tests
```
Comment (ibessonov) [inline modules/file-io/src/test/java/org/apache/ignite/internal/fileio/AbstractFileIoTest.java:563]: Points to the same issue

File: modules/file-io/src/test/java/org/apache/ignite/internal/fileio/FileIoTest.java [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.concurrent.ThreadLocalRandom;
+import org.junit.jupiter.api.Test;
+
+/**
+ * For {@link FileIo} testing.
+ */
+public class FileIoTest {
```
Comment (ibessonov) [inline modules/file-io/src/test/java/org/apache/ignite/internal/fileio/FileIoTest.java:31]: I don't really understand, why do we need this test if there's already AbstractFileIoTest?

Comment (ibessonov) [reply modules/file-io/src/test/java/org/apache/ignite/internal/fileio/FileIoTest.java:31]: I think you should reflect the answer in javadoc so people would understand the difference

File: modules/file-io/pom.xml (hunk 1 above)
Comment (tkalkirill) [reply modules/file-io/pom.xml:62]: Fix it.

File: modules/file-io/src/test/java/org/apache/ignite/internal/fileio/AbstractFileIoTest.java (hunk 5 above)
Comment (tkalkirill) [reply modules/file-io/src/test/java/org/apache/ignite/internal/fileio/AbstractFileIoTest.java:563]: Remove it.

File: modules/file-io/src/test/java/org/apache/ignite/internal/fileio/FileIoTest.java (hunk 6 above)
Comment (tkalkirill) [reply modules/file-io/src/test/java/org/apache/ignite/internal/fileio/FileIoTest.java:31]: At the beginning, I ported it from 2.0, but you're right, it's no longer needed, I'll delete it.

File: modules/file-io/src/test/java/org/apache/ignite/internal/fileio/AbstractFileIoTest.java (hunk 4 above)
Comment (tkalkirill) [reply modules/file-io/src/test/java/org/apache/ignite/internal/fileio/AbstractFileIoTest.java:423]: Here I check only the basic functionality: that there will be no errors when calling **FileIo#force** and that the data will actually be flushed to disk.

File: modules/file-io/src/test/java/org/apache/ignite/internal/fileio/AbstractFileIoTest.java (hunk 2 above)
Comment (tkalkirill) [reply modules/file-io/src/test/java/org/apache/ignite/internal/fileio/AbstractFileIoTest.java:52]: For a simple check that the factory returns the correct instance.

File: modules/file-io/src/test/java/org/apache/ignite/internal/fileio/AbstractFileIoTest.java (hunk 3 above)
Comment (tkalkirill) [reply modules/file-io/src/test/java/org/apache/ignite/internal/fileio/AbstractFileIoTest.java:249]: Modify it

File: modules/file-io/src/test/java/org/apache/ignite/internal/fileio/AbstractFileIoTest.java (hunk 2 above)
Comment (ibessonov) [reply modules/file-io/src/test/java/org/apache/ignite/internal/fileio/AbstractFileIoTest.java:52]: Can you generify this test by introducing method like "abstract Class<? extends FileIoFactory> getIoFactoryClass()"?

Comment (tkalkirill) [reply modules/file-io/src/test/java/org/apache/ignite/internal/fileio/AbstractFileIoTest.java:52]: Fix it.

File: modules/file-io/src/test/java/org/apache/ignite/internal/fileio/AbstractFileIoTest.java (hunk 4 above)
Comment (tkalkirill) [reply modules/file-io/src/test/java/org/apache/ignite/internal/fileio/AbstractFileIoTest.java:423]: Added a comment.

File: modules/file-io/src/test/java/org/apache/ignite/internal/fileio/AbstractFileIoTest.java [hunk 7]
```diff
@@ -560,7 +296,151 @@ void testTransferFrom() throws Exception {
         assertArrayEquals(expectedBytes, toByteArray(test1FilePath));
     }
 
-    // TODO: IGNITE-16988 continue write tests
+    private void checkReadOperation(IoOperation<ByteBuffer, Integer> readOperation) throws Exception {
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/file-io/src/test/java/org/apache/ignite/internal/fileio/AbstractFileIoTest.java:306]: Fix it.

**Pull request merged**